### PR TITLE
Add more livecheck strategies for casks.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

----

This expands support for casks in `brew livecheck` through the following:

* Add support for passing a block after the strategy name in the `livecheck` block DSL. This is primarily to allow us to be able to follow the versioning schemes in casks, where we need to assemble multiple parts of the version into one string. This isn't possible with a `regex` alone, so the strategy block allows for arbitrary logic when necessary. See https://github.com/Homebrew/homebrew-cask/pull/95419 for examples.
* Add a `HeaderMatch` strategy, which is primarily used for checking for versions in `Content-Disposition` and `Location` HTTP headers.
* Add a `Sparkle` strategy, which is used to identify versions from appcast XML data.
* Add `Strategy#page_headers` and `Strategy#page_content` methods. Currently, `#page_headers` is used in `HeaderMatch` and `#page_content` is used in `PageMatch`.
* Add skip conditions for casks to `Livecheck#skip_conditions`. The related statuses are `discontinued`, `latest`, and `unversioned`.

There are some other changes in here but these are the most notable ones.

### CODE REVIEW

File: Library/Homebrew/livecheck/strategy.rb
Comment (samford) [inline Library/Homebrew/livecheck/strategy.rb:84]: Unless we specifically need to use `curl` for retrieving headers, `#page_header` and `#page_contents` can seemingly share one network request, since they're both making `GET` requests to the same URL. As it stands, we're making a minimum of two requests to the same URL and it would multiply further if redirections are involved.

We could seemingly do something like this in both of these methods:

```ruby
URI.parse(url).open do |f|
  @headers = f.meta
  @page_contents = f.read
end
```

In this case, we wouldn't be making duplicate `GET` requests to the same URL when `#page_headers` and `#page_contents` are both used for a given strategy. Calling either of these would store both `@headers` and `@page_contents` and prevent subsequent network requests from being made for the same URL (for a given formula/cask).

One other nice aspect of this is that `OpenURI::Meta` is a hash that's easy to work with and we don't have to worry about parsing the headers:

```ruby
{
  "accept-ranges" => "bytes",
  "age" => "433349",
  "cache-control" => "max-age=604800",
  "content-type" => "text/html; charset=UTF-8",
  "date" => "Sun, 13 Dec 2020 23:36:51 GMT",
  "etag" => "\"3147526947\"",
  "expires" => "Sun, 20 Dec 2020 23:36:51 GMT",
  "last-modified" => "Thu, 17 Oct 2019 07:18:26 GMT",
  "server" => "ECS (dcb/7EEE)",
  "vary" => "Accept-Encoding",
  "x-cache" => "HIT",
  "content-length" => "648"
}
```

To be clear, `@headers` would be the response headers of the last request after any redirections are followed. If you need to get the `location` header before following the redirection, it's necessary to use `.open(redirect: false)` and handle the redirection manually like I'm doing in #9535. I may have overlooked something but it seems like this approach could be adapted to cover all of our needs without needing to duplicate requests.

**Review by samford — COMMENTED** [review]
I know this is a work in progress but I wanted to provide some early feedback regarding a few areas. I haven't gone through all of these changes in detail yet but there are a few important ones that I think we should discuss before we become too attached to a certain implementation.

Of course, please do correct me if I've misunderstood any of this. It's quite a lot to go through and I've been spinning a lot of plates lately, so my brain's more tired than usual.

File: Library/Homebrew/livecheck/strategy/sparkle.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:27]: We should fix the logic to not get to the point where we call this if another strategy with higher priority already matched.

Comment (samford) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:27]: We specifically check all the strategies against a URL (instead of ordering strategies by priority and stopping at the first match) so that we can output the list of matching strategies in the debug output and verbose JSON output.

Knowing which strategies apply to a URL is useful when working on `livecheck` blocks (or analyzing the current state of the formulae/casks in a tap) and I don't feel that the technical requirements of one strategy should require us to drop this longstanding feature.

If we don't agree on this, we should solicit feedback from other maintainers who work with livecheck regularly (i.e., stakeholders) to see what they think. When we've faced decisions like this in the past, we've discussed the benefits and detriments and collectively chosen a path forward.

File: Library/Homebrew/livecheck.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck.rb:71]: See https://github.com/Homebrew/homebrew-cask/pull/95419 for some examples.

File: Library/Homebrew/livecheck/strategy/github_latest.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy/github_latest.rb:40]: > Could you explain how this change is necessary/beneficial?

I don't think it's necessary to do this via the priority. The problem currently is this: Even if a cask explicitly specifies a GitHub `releases.atom` appcast, you would still have to manually specify `:github_latest`.

File: Library/Homebrew/livecheck/strategy.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy.rb:89]: > To be clear, `@headers` would be the headers of the final response,

This will not work. I have already encountered some cases where the first header is a `302` redirect containing the needed `location` header, which is not contained in the final header which only contains e.g. the file name in `content-disposition`.

So currently, all headers are successively merged into one hash containing all headers.

File: Library/Homebrew/rubocops/cask/constants/stanza.rb
Comment (reitermarkus) [reply Library/Homebrew/rubocops/cask/constants/stanza.rb:11]: I'm not sure where it should go. It's basically a replacement for `appcast`, so I thought it should be in the same place.

Another option would be a new group with `[:auto_updates, :livecheck]`.

Comment (vitorgalvao) [reply Library/Homebrew/rubocops/cask/constants/stanza.rb:11]: > My intention behind the previous grouping was to produce something like

That looks better to me. Having a block in the middle of lines (`appcast` position) is hard to follow. I’d even go so far as saying it should be right at the bottom, below everything else, because it’s a block that should remain mostly static (all going well).

Comment (samford) [reply Library/Homebrew/rubocops/cask/constants/stanza.rb:11]: I made [the location/grouping for `livecheck` blocks in casks](https://github.com/Homebrew/brew/blob/b0a3b60292b7919f22a43d40e786ef35562d22df/Library/Homebrew/rubocops/cask/constants/stanza.rb#L12) as close to [the location in formulae](https://github.com/Homebrew/brew/blob/b0a3b60292b7919f22a43d40e786ef35562d22df/Library/Homebrew/rubocops/components_order.rb#L27) as possible and it would be nice to maintain consistency between core/cask. Keeping the `livecheck` block by itself also helps readability.

---

> Another option would be a new group with `[:auto_updates, :livecheck]`.

Currently `livecheck` is in its own group that comes before `auto_updates` and I think that makes sense.

It's not uncommon for us to have comments before the `livecheck` block (explaining an unusual situation). Placing `auto_updates` before `livecheck` in the same group would look weird when there are explanatory comments:

```ruby
auto_updates true
# Aften has moved from a version scheme like 0.07 to 0.0.8. We restrict
# matching to versions with three parts, since a version like 0.07 is parsed
# as 0.7 and seen as newer than 0.0.8.
livecheck do
  url :stable
  regex(%r{url=.*?/aften[._-]v?(\d+(?:\.\d+){2,})\.t}i)
end
```

Placing `auto_updates` after `livecheck` in the same group could lead to `auto_updates` being overlooked. I think the current setup is the most clear and having the cask location in line with formulae helps to keep it predictable between core/cask.

```ruby
# Aften has moved from a version scheme like 0.07 to 0.0.8. We restrict
# matching to versions with three parts, since a version like 0.07 is parsed
# as 0.7 and seen as newer than 0.0.8.
livecheck do
  url :stable
  regex(%r{url=.*?/aften[._-]v?(\d+(?:\.\d+){2,})\.t}i)
end
auto_updates true
```

Additionally, if you group `auto_updates` and `livecheck` together, `auto_updates` appears on its own line when there isn't a `livecheck` block. I think it's best to leave `auto_updates` at the top of the group that comes after `livecheck`. They're already near to one another and grouping them together seems like it would detract from readability rather than improve it.

---

> I’d even go so far as saying it should be right at the bottom, below everything else, because it’s a block that should remain mostly static (all going well).

Part of our thinking when originally selecting a location for the `livecheck` block in formulae was that it would appear near the URLs it uses (`homepage`, `url`/`stable`, `head` for formulae or `url`, `appcast`, `homepage` for casks). This is important because a `livecheck` block should be checked (and updated) when the URLs in a cask/formula notably changes (i.e., something more than a version bump).

Placing the `livecheck` block at the bottom of a cask would distance it from these URLs and make it more likely to be overlooked in this circumstance. Even with the `livecheck` block close to these URLs in homebrew/core formulae, we've already had problems with people changing the `stable` source and not updating the `livecheck` block, so it's better not to make this even easier to miss.

Comment (vitorgalvao) [reply Library/Homebrew/rubocops/cask/constants/stanza.rb:11]: > I made [the location/grouping for `livecheck` blocks in casks](https://github.com/Homebrew/brew/blob/b0a3b60292b7919f22a43d40e786ef35562d22df/Library/Homebrew/rubocops/cask/constants/stanza.rb#L12) as close to [the location in formulae](https://github.com/Homebrew/brew/blob/b0a3b60292b7919f22a43d40e786ef35562d22df/Library/Homebrew/rubocops/components_order.rb#L27) as possible and it would be nice to maintain consistency between core/cask.

Casks and formulae are fundamentally different in how they’re built, so I don’t think that matters and that it’s more important for them to be consistent and easy to parse within themselves.

> This is important because a `livecheck` block should be checked (and updated) when the URLs in a cask/formula notably changes (i.e., something more than a version bump).

But that convinces me.

> Keeping the `livecheck` block by itself also helps readability.

And that’s what I find most important anyway. I don’t think this should be grouped with any other stanza (including `auto_updates`). We don’t do it for any other block; they’re always kept separate.

File: Library/Homebrew/livecheck/strategy/github_latest.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy/github_latest.rb]: Reverted this for now.

File: Library/Homebrew/rubocops/cask/constants/stanza.rb
Comment (reitermarkus) [reply Library/Homebrew/rubocops/cask/constants/stanza.rb]: Reverted this now to the existing location.

**Comment by reitermarkus** [comment]
Merging this now. I did two livechecks of `homebrew/core`, before and after this change, and it seems like all is working apart from some “redirection forbidden” differences which should be handled by https://github.com/Homebrew/brew/pull/9535.

**Comment by samford** [comment]
If you wouldn't mind, please hold off on merging it until I have a chance to review.

**Comment by reitermarkus** [comment]
I'd like to get this in soonish, so we can actually start adding it to casks. Please merge once you reviewed it.

**Review by samford — COMMENTED** [review]
I've left comments below on a few areas to discuss.

Some of the other comments are just explanations of changes that I'll be pushing commits for. I'll resolve the comments that the commits address but I wanted to provide explanations, for reference.

I'm still in the process of reviewing and testing this but it's very late over here and I need to sleep. I'm also going through Homebrew/homebrew-cask#95419, to understand how you're using the `strategy` block (and these new strategies) in practice. I've noticed that a number of those `livecheck` blocks don't follow existing `regex` guidelines, so I would like to take a pass at that PR to bring them up to standard before it's merged as well. Folks who aren't very familiar with livecheck tend to just copy/paste existing `livecheck` blocks when creating a new one, so it's best to nip these issues in the bud.

I know you're eager to work on `livecheck` blocks in homebrew-cask but I want to make sure that we have everything worked out properly beforehand. It's easier to do this now compared to later, so I appreciate your patience. I'm working through this as quickly as I can but I've run out of time/energy for today (it's 2am over here). I'll return to this with fresh eyes first thing in the morning.

File: Library/Homebrew/livecheck/strategy.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy.rb:94]: We can improve this in a follow-up PR.

File: Library/Homebrew/livecheck/strategy.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy.rb]: Yes, it seems like this check isn't necessary anymore since it's explicitly specified in all casks in https://github.com/Homebrew/homebrew-cask/pull/95419 now.

File: Library/Homebrew/livecheck/strategy.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy.rb]: > more flexibility in how we could potentially use `#page_headers` elsewhere in the future

If we need it to be more flexible, we can refactor it in the future when we actually need it.

Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy.rb]: I agree it would be nicer to have all headers separately, but this really shouldn't be a blocker for this PR.

File: Library/Homebrew/livecheck/strategy.rb
Comment (samford) [reply Library/Homebrew/livecheck/strategy.rb:94]: I already have this worked out as part of using `debug: false` for `curl` here and I'm planning to push the changes (along with the others) once the related PRs are merged.

I've stripped the `curl` changes from #10066 and I'll be opening in another PR right after that's merged. The `curl` PR will also include the `retry_request` parameter, so we could use it here. If we don't agree on the idea of the `retry_request` in `curl`, I can simply drop it in favor of only merging the commit to add `debug` to `#curl_with_workarounds`.

Since the work for all this is already done and it's part of some required changes, I don't see any harm in taking care of it now. However, if you think that there's an argument for leaving `--retry` enabled, I'll drop it for now.

File: Library/Homebrew/livecheck/strategy.rb
Comment (samford) [reply Library/Homebrew/livecheck/strategy.rb]: For what it's worth, this refactoring is part of the commits I pushed last night, so it's already done and I only mentioned it to explain the change.

File: Library/Homebrew/livecheck/strategy/bitbucket.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy/bitbucket.rb]: > this doesn't seem to be related to supporting casks

There was at least one cask which was wrongfully using this strategy because the `URL_MATCH_REGEX` here is too broad. I guess it's not an issue anymore since `strategy :sparkle` is now specified explicitly in all casks.

File: Library/Homebrew/livecheck/livecheck.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/livecheck.rb:204]: I agree that for the “Unable to get versions” we don't need a backtrace, for the other ones I would definitely always want the backtrace.

File: Library/Homebrew/livecheck/strategy/bitbucket.rb
Comment (samford) [reply Library/Homebrew/livecheck/strategy/bitbucket.rb]: If I understand this correctly, it sounds like both `Bitbucket` and `Sparkle` matched the URL (when `Sparkle was applied automatically previously) and `Bitbucket` was being applied due to having the higher priority when you wanted to use `Sparkle`.

If that's the case, then specifying the strategy in the `livecheck` block was the appropriate fix in this situation. We have formulae that do this in homebrew/core with `strategy :page_match` when a more specific strategy applies to a given URL (e.g., using `PageMatch` to check a SourceForge page instead of checking the RSS feed using `Sourceforge`).

---

I agree that the existing `URL_MATCH_REGEX` in `Bitbucket` isn't properly enforcing the URL requirements (due to being too loose). I also agree that using the same regex in `#match?` and `#find_versions` is a good idea in a general sense, though it's still frustrating that we can only use `=~` with a regex literal. [I worked on this same refactoring in the past and didn't find a solution for addressing the immutability of named capture group values when using `#match`.]

That said, I think the benefit of using the regex in both places outweighs the detriment of the named capture group values being immutable and we'll be going in this direction anyway. Since this addressed an issue with a cask at one point, I'm fine leaving the `Bitbucket` changes in and I'll just bring the other strategies in line in a follow-up PR afterward.

File: Library/Homebrew/livecheck/livecheck.rb
Comment (samford) [reply Library/Homebrew/livecheck/livecheck.rb]: Could you explain how the backtrace would be used for the other two errors mentioned (`The Sparkle strategy does not support regular expressions.` and `Return value of `strategy :page_match` block must be a string or array of strings.`)?

The debug output already tells us which formula/cask is encountering the issue and whether a `livecheck` block is present (though it will always be "Yes" for these two errors). What does the backtrace provide that the debug output and error message aren't already covering?

Comment (reitermarkus) [reply Library/Homebrew/livecheck/livecheck.rb]: Basically, the other two are actually errors which should never happen anyways, so in practice you will not see the backtrace unless something is actually wrong. I agree that the backtrace doesn't really add anything useful for these two, but I also don't think it's worth the special handling when they shouldn't occur in the first place.

**Comment by reitermarkus** [comment]
@samford, any blockers left here? I think this should be good to go now.

File: Library/Homebrew/livecheck/livecheck.rb
Comment (samford) [reply Library/Homebrew/livecheck/livecheck.rb]: What's notable about these two errors is that they occur because of misconfiguration in a `livecheck` block, not a bug in `brew livecheck`. We only encounter these two errors when a maintainer/contributor has done something they shouldn't have in a `livecheck` block and I think the error sufficiently explains what they did wrong (i.e., what they need to change).

Can you explain what the backtrace would be providing in the context of these two error messages that the debug output and error message don't already cover?

**Comment by samford** [comment]
> any blockers left here?

I have some commits to push to this PR branch after #10069 is merged but I believe that should be the last big thing. I'm going to rebase this on #10069 locally and run some final tests to make sure we haven't broken anything along the way.

Once #10069 is merged and I've finished testing, I'll rebase this on `master` and push my latest commits. There may be a little discussion around those changes (nothing too notable) but I agree that this is pretty much done at this point. My commits address will address some remaining areas where I have concerns, so that should finish things off. Any other areas aren't notable enough to hold this up and we can address them in follow-up PRs.

I believe the only other thing is finishing off the backtrace discussion. I'm not convinced the backtrace is beneficial for those other two errors, so I would like to better understand your reasoning.

All that said, if we can get #10069 merged, I'll wrap things up in this PR and merge it once it's ready.

File: Library/Homebrew/livecheck/livecheck.rb
Comment (reitermarkus) [reply Library/Homebrew/livecheck/livecheck.rb]: > what the backtrace would be providing

As I already mentioned, I agree that the backtrace doesn't provide any additional meaningful information. My point is that in my opinion the special treatment of this is not worth the effort/complexity. If you want to add the logic for ignoring these two errors specifically, feel free to do so in a follow-up PR.

Comment (reitermarkus) [reply Library/Homebrew/livecheck/livecheck.rb]: In either case, whether or not there is a backtrace, the error should be fixed in the formula/cask causing it, so in that regard the backtrace might even provide some kind of incentive to do just that.

Comment (samford) [reply Library/Homebrew/livecheck/livecheck.rb]: I agree that this is a minor detail and it's going to be rare for anyone to run into these errors. `Unable to get versions` was the notable one to omit the backtrace on and you've handled that. I might revisit this in the future if I find the time but, with this discussion in mind, I'm fine with where this is. Thanks for the explanation.

**Comment by samford** [comment]
I've rebased this on `master` now that #10069 is merged. I'm doing a couple final runs of livecheck to compare output but I went ahead and pushed my latest commits so you can have a look at the changes in the interim time. These are the last notable changes that I anticipate unless I encounter any issues during testing. If you sign off on these changes, then I'll take care of wrapping this up and merging it when it's ready to go.

While livecheck is running in the background, I'm working on adding more tests to bump up the code coverage to a passing level. I may also take a pass through this to add documentation comments but these won't really require review. I'm sure it's late over there and I don't want to keep you up while I work on tying up loose ends.

File: Library/Homebrew/livecheck/strategy.rb
Comment (samford) [inline Library/Homebrew/livecheck/strategy.rb:93]:
```suggestion
```

We don't need `--fail` here because it's already included by `#curl_args` in this scenario.

Comment (reitermarkus) [reply Library/Homebrew/livecheck/strategy.rb]: Right, I just noticed that, removed the commit again.

Comment (samford) [reply Library/Homebrew/livecheck/strategy.rb]: Feel free to push the other changes that were part of that commit, though. I had forgotten that you can do `stdout, _, status = curl_with_workarounds(...)`.

**Review by samford — APPROVED** [review]
I wrote tests for what I could but most of livecheck can't be tested without making network requests. Strategy tests that involve network requests were very unreliable on CI when we originally tried them out (i.e., inexplicably failing on CI though they worked fine locally), so we continue to avoid them until/unless something changes. We're also not allowed to stub out the network requests (as an alternative), so we can't test a lot of livecheck's code in this state.

Expanding the `livecheck/livecheck.rb` tests wasn't enough to get code coverage to pass. The only other non-networked code I could find to test was `Sparkle#item_from_content`, so I made public for the sake of being able to test it. This was the only private method in a livecheck strategy, so making it public simply brings it in line with the others.

If memory serves, I believe I originally made the internal methods in strategies (e.g., `Git#tag_info`) private when I refactored the codebase into its current structure. However, keeping these methods private was deemed unnecessary when livecheck was going through review as part of the migration from Homebrew/homebrew-livecheck to Homebrew/brew. On the one hand, I think it makes sense for them to be private but, on the other hand, it doesn't seem like we can test private methods.

At the very least, it was necessary to make `#item_from_content` public for testing to get enough code coverage for this PR to pass CI so we can merge it, so it is what it is.

---

I've done comparisons of verbose JSON livecheck output between `master` and this branch, across both homebrew/core and homebrew/cask, and nothing jumped out at me. When testing homebrew/cask, I also used the branch from Homebrew/homebrew-cask#95419, which includes the WIP livecheck blocks.

There's potentially more room for improvement here but I'm good with the current state of this. It's passing CI now and we've addressed everything important, so I think it's fine to merge. We can always improve on this in follow-up PRs but I don't see any reason to hold this up further.

---

Looking to the future, I'm planning to go through Homebrew/homebrew-cask#95419 and bring it up to our existing standards. Namely, there are a number of regexes that need to be brought up to standard. Since we're introducing `strategy` blocks in this PR, I would also like to review the usage in the homebrew/cask PR to make sure we're providing good examples early on.

I'm going to try to work through that PR tomorrow, as I don't want to unnecessarily hold you up from moving it forward. I do think it's important to address it now rather than later, as it's comparatively easy to review/fix in that PR rather than having to hunt down each individual `livecheck` block after it's merged.

---

Thanks, @reitermarkus!

**Pull request merged**
